controller/multi_writer_at.go

```diff
@@ -110,3 +110,91 @@ func (m *MultiWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
 	}
 	return len(p), nil
 }
+
+func (m *MultiWriterAt) Sync() (err error) {
+	replicaErrs := make([]error, len(m.writers))
+	replicaErrCount := 0
+	replicaErrored := false
+	wg := sync.WaitGroup{}
+	var errors MultiWriterError
+	var multiWriterMtx sync.Mutex
+
+	for i, w := range m.writers {
+		wg.Add(1)
+		go func(index int, w Writer) {
+			err := w.Sync()
+			if err != nil {
+				multiWriterMtx.Lock()
+				replicaErrored = true
+				replicaErrs[index] = err
+				multiWriterMtx.Unlock()
+			}
+			wg.Done()
+		}(i, w)
+	}
+	wg.Wait()
+	if replicaErrored {
+		errors.Writers = m.writers
+		errors.ReplicaErrors = replicaErrs
+	}
+	//Below code is introduced to make sure that the IO has been written to more
+	//than 50% of the replica.
+	//If any replica has errored, return with the length of data written and the
+	//erroed replica details so that it can be closed.
+	if replicaErrored {
+		for _, err1 := range replicaErrs {
+			if err1 != nil {
+				replicaErrCount++
+			}
+		}
+		if len(m.writers)-replicaErrCount > len(m.writers)/2 {
+			return &errors
```

Comment:
better to have different return values in case of partial success for 'having required success_cnt' and 'not having required success_cnt'